apis/network/v1beta1/netconfig_types.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:validation:Pattern="^[a-zA-Z0-9][a-zA-Z0-9\\-_]*[a-zA-Z0-9]$"
+// NetNameStr is used for validation of a net name.
+type NetNameStr string
+
+// +kubebuilder:validation:Pattern="^[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9]$"
+// HostNameStr is used for validation of a hostname.
+type HostNameStr string
+
+// Network definition
+type Network struct {
+	// +kubebuilder:validation:Required
+	// Name of the network, e.g. External, InternalApi, ...
+	Name NetNameStr `json:"name"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=1500
+	// MTU of the network
+	MTU int `json:"mtu"`
+
+	// +kubebuilder:validation:Required
+	// Subnets of the tripleo network
+	Subnets []Subnet `json:"subnets"`
+}
+
+// Subnet definition
+type Subnet struct {
+	// +kubebuilder:validation:Required
+	// Name of the subnet
+	Name NetNameStr `json:"name"`
+
+	// +kubebuilder:validation:Required
+	// Cidr the cidr to use for this network
+	// TODO: add validation to webhook
+	Cidr string `json:"cidr"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Maximum=4094
+	// Vlan ID
+	Vlan int `json:"vlan,omitempty"`
+
+	// +kubebuilder:validation:Required
+	// AllocationRanges a list of AllocationRange for assignment. Allocation will start
+	// from first range, first address.
+	AllocationRanges []AllocationRange `json:"allocationRanges"`
+
+	// +kubebuilder:validation:Optional
+	// Gateway optional gateway for the network
+	// TODO: add validation to webhook
+	Gateway *string `json:"gateway,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// Routes, list of networks that should be routed via network gateway.
+	Routes []Route `json:"routes,omitempty"`
+}
+
+// AllocationRange definition
+type AllocationRange struct {
+	// +kubebuilder:validation:Required
+	// Start a set of IPs that are reserved and will not be assigned
+	// TODO: add validation to webhook
+	Start string `json:"start"`
+
+	// +kubebuilder:validation:Required
+	// End a set of IPs that are reserved and will not be assigned
+	// TODO: add validation to webhook
+	End string `json:"end"`
+
+	// +kubebuilder:validation:Optional
+	// ExcludeAddresses a set of IPs that should be excluded from range
+	// TODO: add validation to webhook
+	ExcludeAddresses []string `json:"excludeAddresses,omitempty"`
```

Comment:
I know I asked for exclude's in allocation pool earlier, but when I look at it again now I think it would it be better to move the ExcludeAddresses to the subnet struct? Since we use it both for in allocation range and outside allocation range in fixedIPs case, feels like the UX is better if this is per subnet - not per-allocation range.

If we set it at Subnet, and place in the AssignIpDetails struct? Code in FixedIPs would become simpler as well since no need to iterate AllocationRanges.

Reply:
ack sounds good

Reply:
done